File: python/day-03/part1.py
grid = open(0).read().splitlines()
coords = set()

for r, row in enumerate(grid):
    for c, ch in enumerate(row):
        if ch.isdigit() or ch == ".":
            continue

        for cr in [r - 1, r, r + 1]:
            for cc in [c - 1, c, c + 1]:
                if not 0 <= cr < len(grid) or not 0 <= cc < len(row) or not grid[cr][cc].isdigit():
                    continue

                while cc > 0 and grid[cr][cc - 1].isdigit():
                    cc -= 1

                coords.add((cr, cc))

s = 0
for r, cs in coords:
    ce = cs + 1
    while ce < len(grid[0]) and grid[r][ce].isdigit():
        ce += 1
    s += int(grid[r][cs:ce])

print(s)

File: rust/day-03/src/part1.rs
use std::collections::HashSet;

fn checked_add_signed_bounded(a: usize, b: isize, bound: usize) -> Option<usize> {
    match a.checked_add_signed(b) {
        Some(s) if s < bound => Some(s),
        _ => None,
    }
}

pub fn solve(input: &str) -> u32 {
    let grid = input
        .lines()
        .map(|line| line.chars().collect::<Vec<_>>())
        .collect::<Vec<_>>();

    let mut parts: HashSet<[usize; 2]> = HashSet::new();

    for (r, row) in grid.iter().enumerate() {
        for (c, &ch) in row.iter().enumerate() {
            if ch == '.' || ch.is_digit(10) {
                continue;
            }

            #[rustfmt::skip]
            (-1..=1).flat_map(|dr| (-1..=1).map(move |dc| [dr, dc]))
                .filter_map(|[dr, dc]| {
                    checked_add_signed_bounded(r, dr, grid.len())
                        .zip(
                    checked_add_signed_bounded(c, dc,  row.len()))})
                .for_each(|(cr, mut cc)| {
                    if !grid[cr][cc].is_digit(10) {
                        return;
                    }

                    while cc > 0 && grid[cr][cc - 1].is_digit(10) {
                        cc -= 1;
                    }

                    parts.insert([cr, cc]);
                });
        }
    }

    parts
        .iter()
        .map(|&[r, cs]| {
            let mut ce = cs + 1;

            while ce < grid[0].len() && grid[r][ce].is_digit(10) {
                ce += 1;
            }

            grid[r][cs..ce]
                .iter()
                .collect::<String>()
                .parse::<u32>()
                .unwrap()
        })
        .sum()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solve() {
        let input = std::fs::read_to_string("inputs/test.txt")
            .expect("ERROR: inputs/test.txt file not found");

        assert_eq!(solve(&input), 4361);
    }
}
